src/shared/utilities/messages.ts

```diff
@@ -182,11 +182,134 @@ export async function showMessageWithCancel(
+        super(timeoutLength)
+    }
+
+    public override cancel(): void {
+        super.cancel()
+        CancellableMessageManager.cancelMessage(this.msgId)
+    }
+
+    public override dispose(): void {
+        super.dispose()
+        CancellableMessageManager.disposeMessage(this.msgId)
+    }
+}
+
+/**
+ * Provides support to create cancellable vscode message windows.
+ *
+ * To get started, you create a new message with a unique ID.
+ * Following that, you can easily update the message or cancel it.
+ *
+ * ---
+ *
+ * This class solves an issue of needing to manage duplicated
+ * messages, the interface simplifies the process.
+ */
+export class CancellableMessageManager {
+    static readonly timeoutMillis = 60000
+    private static messageMap: {
+        [msgId: string]: { timeout: CancellableMessageTimeout; progress: vscode.Progress<ProgressEntry> }
+    } = {}
+
+    /**
+     * Starts a message if it does not exist, then applies the update with the given
+     * progress data.
+     */
+    static async putMessage(
+        msgId: string,
+        messageText: string,
+        progressEntry?: ProgressEntry,
+        timeoutMillis?: number
+    ): Promise<void> {
+        let message = this.messageMap[msgId]
+
+        if (message === undefined) {
+            await this.startMessage(msgId, messageText, timeoutMillis)
+            message = this.messageMap[msgId]
+        }
+
+        if (progressEntry !== undefined) {
+            message.progress.report(progressEntry)
+        }
+    }
+
+    /**
+     * @param timeoutMillis The amount of milliseconds till the message cancels itself.
+     */
+    static async startMessage(
+        msgId: string,
+        messageText: string,
+        timeoutMillis: number = this.timeoutMillis
+    ): Promise<void> {
+        this.cancelMessage(msgId) // cancels if exists
+        const timeout = new CancellableMessageTimeout(timeoutMillis, msgId)
+        this.messageMap[msgId] = { timeout, progress: await showMessageWithCancel(messageText, timeout) }
+    }
+
+    static getTimeout(msgId: string): CancellableMessageTimeout {
+        const message = this.messageMap[msgId]
+        if (message === undefined) {
+            throw Error(`Cannot get timeout for '${msgId}', it does not exist.`)
+        }
+        return message.timeout
+    }
+
+    /**
+     * Updates an EXISTING message
+     */
+    static updateMessage(msgId: string, progressUpdate: ProgressEntry): void {
+        const message = this.messageMap[msgId]
+
+        if (message === undefined) {
+            throw Error(`Cannot update progress for '${msgId}', it does not exist.`)
+        }
+
+        message.progress.report(progressUpdate)
+    }
+
+    /** Cancels the message and clears it from the manager. */
+    static cancelMessage(msgId: string): void {
+        this.stop(msgId, 'cancel')
+    }
+
+    /** Disposes the message and clears it from the manager. */
+    static disposeMessage(msgId: string): void {
```

Comment:
could eliminate this, cancelMessage, and getTimeout by instead exposing `getMessage`.